### PR TITLE
Updated notebook about PRs and reviewing

### PR DESCRIPTION
**Pull request by astrofrog**
@eteq and @bsipocz - can you take a look? I'm going to work on the actual code/docs in the example repo now.

### CODE REVIEW

File: day2_software/prs_and_reviewing.ipynb [hunk 1]
```diff
@@ -9,241 +9,161 @@
...
-   "source": [
-    "Now if you look at the local repo, it should include your changes."
-   ]
-  },
-  {
-   "cell_type": "markdown",
-   "metadata": {},
-   "source": [
-    "*Suggestion* You mauy want to change the \"origin\" remote to your username.  E.g. ``git remote rename origin <yourusername>``.  To go further, you might even *delete* your fork's `master` branch, so that only your neighbor's `master` exists.  That might save you headaches in the long run if you were to ever access this repo again in the future. "
+    "* It allows other people to review the changes you are suggesting and iterate on them before merging\n",
```
Comment (bsipocz) [inline day2_software/prs_and_reviewing.ipynb]:
```suggestion
    "* It allows other people to review the changes you are suggesting and iterate on them before actually changing the code.\n",
```

Comment (bsipocz) [reply day2_software/prs_and_reviewing.ipynb]: just to avoid calling it "merging" not a hard preference

File: day2_software/prs_and_reviewing.ipynb [hunk 2]
```diff
@@ -9,241 +9,161 @@
...
+    "* It also allows anyone to propose contributions to any repository without taking on the risk of making many people have write-access to the main repository\n",
```
Comment (bsipocz) [inline day2_software/prs_and_reviewing.ipynb]:
```suggestion
    "* It also allows anyone to contribute by proposing changes to any repository without taking on the risk of making many people have write-access to the main repository.\n",
```

File: day2_software/prs_and_reviewing.ipynb [hunk 3]
```diff
@@ -9,241 +9,161 @@
...
+    "    \n",
+    "You can then switch to this branch with:\n",
+    "\n",
+    "    git checkout <name-of-branch>\n",
+    "    \n",
+    "Note that you can also write ``git checkout -b <name-of-branch>`` as a short-hand for the above two commands. Branch names should be kept simple and reasonably short, and can't include e.g. spaces.\n",
+    "\n",
+    "### 1d: Make some changes\n",
+    "\n",
+    "At this point, you can pick what kind of changes you would like to make - these are ordered by increasing difficulty. You can choose to tackle more than one of these if you like, but since we are limited in time, the aim is to just do one of these:\n",
```
Comment (bsipocz) [inline day2_software/prs_and_reviewing.ipynb]: maybe rephrase "increasing difficulty"? I don't have a better suggestion though.

File: day2_software/prs_and_reviewing.ipynb [hunk 4]
```diff
@@ -163,7 +163,37 @@
    "source": [
     "## Problem 3: Making additional changes\n",
     "\n",
-    "If someone comments on your pull request and you decide to follow their recommendations, you can add new changes to the pull request without having to open a new one. To do this, just make any changes you need, then stage the files to change with ``git add`` as described above, then commit the changes with ``git commit`` and use ``git push`` to push the changes to your branch. The commit will then appear in the pull request."
+    "If someone comments on your pull request and you decide to follow their recommendations, you can add new changes to the pull request without having to open a new one. To do this, just make any changes you need, then stage the files to change with ``git add`` as described above, then commit the changes with ``git commit`` and use ``git push`` to push the changes to your branch. The commit will then appear in the pull request.\n",
+    "\n",
+    "## Next steps: Documentation or Tests - Choose your own adventure!\n",
```
Comment (bsipocz) [inline day2_software/prs_and_reviewing.ipynb]: you can call this Problem 4, to follow the style of the previous exercises

File: day2_software/prs_and_reviewing.ipynb [hunk 5]
```diff
@@ -163,7 +163,37 @@
...
+    "\n",
+    "At this point, there are two things you can learn more about:\n",
+    "\n",
+    "* While a README file is fine for very simple code, in general we recommend that you write more fully-fledged documentation using a framework such as [Sphinx](http://www.sphinx-doc.org/en/master/). If you want to learn about how to do this, you can follow the notebook in this folder on **Code Documentation**.\n",
```
Comment (bsipocz) [inline day2_software/prs_and_reviewing.ipynb]: shall we link to the other notebooks? No strong preference either way.

File: day2_software/prs_and_reviewing.ipynb [hunk 6]
```diff
@@ -9,241 +9,189 @@
...
-   "source": [
-    "### 2c: create a branch for your change"
-   ]
-  },
-  {
-   "cell_type": "markdown",
-   "metadata": {},
-   "source": [
-    "You're going to make some changes to their code, but who knows... maybe they'll spend so long reviewing it that you want to do another.  So it's always best to make changes in a specific \"branch\" for that change.  So to do this we need to make a github branch.\n",
+    "Whether you are working as part of a collaboration that is using one or more git repositories for common code, or whether you want to contribute to a package that is being developed in a git repository, a very useful skill is that of being able to make contributions to these repositories using *pull requests* (as they are called on GitHub - they are called e.g. *merge requests* on GitLab).\n",
```
Comment (eteq) [inline day2_software/prs_and_reviewing.ipynb]:
```suggestion
    "A very useful skill in both astronomy and the wider software world to be able to make contributions to software repositories using *pull requests*. (This is what they are called on GitHub - they go by other similar names in other places - e.g. *merge requests* on GitLab).  This is useful whether you are working as part of a collaboration that is using git repositories for common code, or if you want to contribute to an open code base that is being developed in a git repository, \n",
```

I felt this "buried the lead" (and was a bit run-on) so I tried to invert the ordering of the sentence while leaving the content otherwise similar.

File: day2_software/prs_and_reviewing.ipynb [hunk 7]
```diff
@@ -9,241 +9,189 @@
...
-   "source": [
-    "### 2c: modify the code"
-   ]
-  },
-  {
-   "cell_type": "markdown",
-   "metadata": {},
-   "source": [
-    "Make some change to their code repo.  Usually this would be a new feature or a bug fix or documentation clarification or the like... But it's up to you.  \n",
+    "There are two main benefits of making contributions via pull requests:\n",
```
Comment (eteq) [inline day2_software/prs_and_reviewing.ipynb]:
```suggestion
    "There are two main benefits to making contributions via pull requests:\n",
```

File: day2_software/prs_and_reviewing.ipynb [hunk 8]
```diff
@@ -9,241 +9,189 @@
...
-   "source": [
-    "Now if you look at the local repo, it should include your changes."
-   ]
-  },
-  {
-   "cell_type": "markdown",
-   "metadata": {},
-   "source": [
-    "*Suggestion* You mauy want to change the \"origin\" remote to your username.  E.g. ``git remote rename origin <yourusername>``.  To go further, you might even *delete* your fork's `master` branch, so that only your neighbor's `master` exists.  That might save you headaches in the long run if you were to ever access this repo again in the future. "
+    "* It allows other people to review the changes you are suggesting and iterate on them before they are merged (i.e. before they are incorporated into the main repository)\n",
```
Comment (eteq) [inline day2_software/prs_and_reviewing.ipynb]:
```suggestion
    "* It allows other people to review the changes you are requesting and make their own suggestions on your code before they are merged (i.e. before they are incorporated into the main repository)\n",
```

File: day2_software/prs_and_reviewing.ipynb [hunk 9]
```diff
@@ -9,241 +9,189 @@
...
+    "* It also allows anyone to contribute by proposing changes to any repository without taking on the risk of making many people have write-access to the main repository.\n",
```
Comment (eteq) [inline day2_software/prs_and_reviewing.ipynb]:
```suggestion
    "* It allows anyone to contribute by proposing changes to any repository without taking on the risk of many people having write-access to the main repository.\n",
```

File: day2_software/prs_and_reviewing.ipynb [hunk 10]
```diff
@@ -9,241 +9,189 @@
...
+    "\n",
+    "In this tutorial, we will go through the process of opening a pull request, and we will also take some time to review other participant's pull requests. We will take the example of a repository on GitHub, but note that the workflow will be similar for other platforms such as GitLab and BitBucket.\n",
+    "\n",
+    "## Problem 1: Making some changes and opening a pull request\n",
+    "\n",
+    "\n",
+    "### 1a: Fork the repository\n",
+    "\n",
+    "The repository we will be opening pull requests to is at https://github.com/AstroHackWeek/pr_review_tutorial. Since you don't have write access to this repository, the first step will be to make a 'fork' of the repository. Make sure you are logged in to GitHub, then go to the repository and click the 'Fork' button:\n",
```
Comment (eteq) [inline day2_software/prs_and_reviewing.ipynb]:
```suggestion
    "The repository we will be opening pull requests to is at https://github.com/AstroHackWeek/pr_review_tutorial. Since you don't have write access to this repository, the first step will be to make a "fork" of the repository. This is basically a copy of the main repository, but one that you have complete control to modify as you see fit (and that is visible to everyone else on GitHub).  Make sure you are logged in to GitHub, then go to the repository and click the 'Fork' button:\n",
```

File: day2_software/prs_and_reviewing.ipynb [hunk 11]
```diff
@@ -9,241 +9,189 @@
...
+    "\n",
+    "This will create a copy of the repository at ``https://github.com/<username>/pr_review_tutorial`` where ``<username>`` is your GitHub username.\n",
+    "\n",
+    "### 1b: Clone your fork\n",
+    "\n",
+    "Now that you have forked your repository, you can clone it to your computer using:\n",
+    "\n",
+    "    git clone https://github.com/<username>/pr_review_tutorial\n",
+    "    \n",
+    "Be sure to replace ``<username>`` with your GitHub username! Note that if you are using Windows, you might find it helpful to use [git bash](https://gitforwindows.org/).\n",
```
Comment (eteq) [inline day2_software/prs_and_reviewing.ipynb]: Perhaps the mention of git bash should be at the top?  You might also mention windows subsystem for linux - https://docs.microsoft.com/en-us/windows/wsl/about - which is more "feature complete" in a lot of ways for this workflow

File: day2_software/prs_and_reviewing.ipynb [hunk 12]
```diff
@@ -9,241 +9,189 @@
...
+    "\n",
+    "* ``simple_functions.py``: a set of simple functions that should work properly\n",
+    "\n",
+    "* ``buggy_function.py``: a function that contains one or more bugs\n",
+    "\n",
+    "Take a look at these files, and familiarize yourself with the contents.\n",
+    "\n",
+    "### 1d: Make a new branch\n",
+    "\n",
+    "By default, you will be looking at the ``master`` (or main) \"branch\" of the repository. You're going to make some changes to the code, but who knows... maybe the reviewers will spend so long reviewing your changes that you want to do another set of unrelated changes in the mean time. So it's always best to make changes in a dedicated branch. So before you make any changes, make a new branch with:\n",
```
Comment (eteq) [inline day2_software/prs_and_reviewing.ipynb]:
```suggestion
    "By default, you will be looking at the ``master`` (or main) \"branch\" of the repository. You're going to make some changes to the code, but who knows... maybe the reviewers will spend so long reviewing your changes that you want to do another set of unrelated changes in the mean time. Thus it's always best to make changes in a dedicated branch. So before you make any changes, make a new branch with:\n",
```

File: day2_software/prs_and_reviewing.ipynb [hunk 13]
```diff
@@ -9,241 +9,189 @@
...
+    "\n",
+    "    git branch <name-of-branch>\n",
+    "    \n",
+    "You can then switch to this branch with:\n",
+    "\n",
+    "    git checkout <name-of-branch>\n",
+    "    \n",
+    "Note that you can also write ``git checkout -b <name-of-branch>`` as a short-hand for the above two commands. Branch names should be kept simple and reasonably short, and can't include e.g. spaces.\n",
+    "\n",
+    "### 1d: Make some changes\n",
```
Comment (eteq) [inline day2_software/prs_and_reviewing.ipynb]: I think this is supposed to be 1e, but then all the other ones following need different lettering...

File: day2_software/prs_and_reviewing.ipynb [hunk 14]
```diff
@@ -9,241 +9,189 @@
...
+    "At this point, go to your fork on GitHub at ``https://github.com/<username>/pr_review_tutorial``, and you should see a yellow-ish banner that looks like this:\n",
+    "\n",
+    "<img src=\"images/pushed.png\" width=\"700\" alt=\"The yellow-ish banner on GitHub showing a branch has been pushed\">\n",
+    "\n",
+    "Click on **Compare & pull request**. You will then see a screen like:\n",
+    "\n",
+    "<img src=\"images/compare.png\" width=\"700\" alt=\"A preview of the changes for the pull request\">\n",
+    "\n",
+    "At this point, give your pull request a sensible title, and use the description box to describe what changes you've made, why you made them, and you can also mention if there are any unresolved issues for example. You can also preview the changes by scrolling down on the page.\n",
+    "\n",
```
Comment (eteq) [inline day2_software/prs_and_reviewing.ipynb:133]:
```suggestion
    "\n",
    "If for some reason you don't see the yellow banner (for example, if you went and did something else after pushing so that it timed out), you can also make a pull request by choosing your branch from the drop-down menu and then choosing 'New pull request'\n",
    "\n",
```

File: day2_software/prs_and_reviewing.ipynb [hunk 15]
```diff
@@ -9,241 +9,189 @@
...
+    "<img src=\"images/plus.png\" width=\"300\" alt=\"Click plus for adding comment\">\n",
+    "\n",
+    "At this point, you can make a comment then click **Start a review**. If you add more comments, you will need to click on **Add to review**. Once you've finished adding comments, you can click on **Finish your review** in the top right above the detailed changes, and leave a general message and your recommendation (approve or request changes).\n",
+    "\n",
+    "But what should you comment on? Anything you think can be improved or that you don't understand! For example:\n",
+    "\n",
+    "* You can ask for comments to be added to code that's not intuitive\n",
+    "* You can comment on the coding style\n",
+    "* You can point out mistakes/bugs\n",
```
Comment (eteq) [inline day2_software/prs_and_reviewing.ipynb:157]:
```suggestion
    "* You can point out mistakes/bugs\n",
    "* You can suggest additions to the code using the 'Insert a suggestion' option in the GitHub interface\n",
```

File: day2_software/prs_and_reviewing.ipynb [hunk 16]
```diff
@@ -9,241 +9,189 @@
...
+    "\n",
+    "In general make sure you comment constructively! You can also choose to leave comments that don't require changes, e.g. 'Nice approach!'. If a pull request is perfect, you can just skip adding in-line comments and just leave a general comment and approve the pull request."
```
Comment (eteq) [inline day2_software/prs_and_reviewing.ipynb]:
```suggestion
    "In general make sure you comment constructively (and maybe even friendly)! You can also choose to leave comments that don't require changes, e.g. 'Nice approach!'. If a pull request is perfect, you can just skip adding in-line comments and just leave a general comment and approve the pull request. In general you have the choice to approve or ask for revision - while in a repo like this where you don't have write access, these are not binding, but they are critical signals to the actual maintainers whether you feel the code is ready to be accepted or not.  If you don't feel ready to say something firm either way you can always just leave your review as a comment, which serves to neither ask for changes nor accept the pull request."
```

File: day2_software/prs_and_reviewing.ipynb [hunk 17]
```diff
@@ -9,241 +9,189 @@
...
    "metadata": {},
    "source": [
-    "Science (Data or otherwise) and open source code is a social enterprise built on shared effort, mutual respect, and trust.  So ask them to issue a PR aginst *your* code, too.  The more we can stand on each others' shoulders, the farther we will all see.\n",
+    "## Problem 3: Making additional changes\n",
+    "\n",
+    "If someone comments on your pull request and you decide to follow their recommendations, you can add new changes to the pull request without having to open a new one. To do this, just make any changes you need, then stage the files to change with ``git add`` as described above, then commit the changes with ``git commit`` and use ``git push`` to push the changes to your branch. The commit will then appear in the pull request.\n",
+    "\n",
+    "## Problem 4: Documentation or Tests - Choose your own adventure!\n",
+    "\n",
+    "At this point, there are two things you can learn more about:\n",
```
Comment (eteq) [inline day2_software/prs_and_reviewing.ipynb]:
```suggestion
    "At this point, there are three things you can learn more about, building on the three examples of improvable code that have been outlined above:\n",
```

(goes with the follow-on suggestion)

File: day2_software/prs_and_reviewing.ipynb [hunk 18]
```diff
@@ -9,241 +9,189 @@
...
+    "\n",
+    "* While a README file is fine for very simple code, in general we recommend that you write more fully-fledged documentation using a framework such as [Sphinx](http://www.sphinx-doc.org/en/master/). If you want to learn about how to do this, you can follow the notebook in this folder on **[Code Documentation](documentation.ipynb)**.\n",
+    "\n",
+    "* If you fix bugs or add new functions in code, you really should be adding tests to make sure that the code works and that bugs that have been fixed are not re-introduced. If you want to learn about this, take a look at the **Code [Testing and CI](testing_and_CI.ipynb)** notebook.\n",
+    "\n",
```
Comment (eteq) [inline day2_software/prs_and_reviewing.ipynb:178]:
```suggestion
    "\n",
    "* If you want to know how to set a Python package up from scratch and publish it so that it can be installed from `pip` and/or used by you in *other* places other than where the package itself is located, take a look at the **[Packaging](packaging.ipynb)** notebook.\n",
    "\n",
```

File: day2_software/prs_and_reviewing.ipynb [hunk 19]
```diff
@@ -9,241 +9,189 @@
...
    ]
   },
   {
    "cell_type": "markdown",
    "metadata": {},
    "source": [
-    "Science (Data or otherwise) and open source code is a social enterprise built on shared effort, mutual respect, and trust.  So ask them to issue a PR aginst *your* code, too.  The more we can stand on each others' shoulders, the farther we will all see.\n",
+    "## Problem 3: Making additional changes\n",
+    "\n",
+    "If someone comments on your pull request and you decide to follow their recommendations, you can add new changes to the pull request without having to open a new one. To do this, just make any changes you need, then stage the files to change with ``git add`` as described above, then commit the changes with ``git commit`` and use ``git push`` to push the changes to your branch. The commit will then appear in the pull request.\n",
```
Comment (eteq) [inline day2_software/prs_and_reviewing.ipynb]:
```suggestion
    "If someone comments on your pull request and you decide to follow their recommendations, you can add new changes to the pull request without having to open a new one. To do this, just make any changes you need, then stage the files to change with ``git add`` as described above, then commit the changes with ``git commit`` and use ``git push`` to push the changes to your branch. The commit will then appear in the pull request. Once you think it's all ready, you might want to ask one of the instructors if they are willing to merge your PR.\n",
```

**Review by eteq — CHANGES_REQUESTED** [review]
I left some comments.  I am trying to come up with a rude thing to say to be ironic given one of the suggestions is to be friendly, but nothing is coming to mind.  Maybe ✌️ ?

File: day2_software/prs_and_reviewing.ipynb (hunk 13 above)
Comment (astrofrog) [reply day2_software/prs_and_reviewing.ipynb]: Fixed

File: day2_software/prs_and_reviewing.ipynb (hunk 11 above)
Comment (astrofrog) [reply day2_software/prs_and_reviewing.ipynb]: I've moved this to the top and also mention the Windows Subsystem for Linux too now.

**Comment by astrofrog** [comment]
@eteq - I think I've accepted all your suggestions and implemented the comments. Feel free to accept/merge once you feel it's ready!

**Review by eteq — APPROVED** [review]
